pkg/controller/infrastructure/actuator_delete.go

```diff
@@ -64,9 +68,65 @@ func (a *actuator) Delete(ctx context.Context, log logr.Logger, infra *extension
 		return err
 	}
 
+	openstackClient, err := openstackclient.NewOpenstackClientFromCredentials(credentials)
+	if err != nil {
+		return err
+	}
+
+	networkingClient, err := openstackClient.Networking()
+	if err != nil {
+		return err
+	}
+
 	stateInitializer := terraformer.StateConfigMapInitializerFunc(terraformer.CreateState)
-	return tf.
-		InitializeWith(ctx, terraformer.DefaultInitializer(a.Client(), terraformFiles.Main, terraformFiles.Variables, terraformFiles.TFVars, stateInitializer)).
-		SetEnvVars(internal.TerraformerEnvVars(infra.Spec.SecretRef, credentials)...).
-		Destroy(ctx)
+	tf = tf.InitializeWith(ctx, terraformer.DefaultInitializer(a.Client(), terraformFiles.Main, terraformFiles.Variables, terraformFiles.TFVars, stateInitializer)).SetEnvVars(internal.TerraformerEnvVars(infra.Spec.SecretRef, credentials)...)
+
+	configExists, err := tf.ConfigExists(ctx)
+	if err != nil {
+		return err
+	}
+
+	vars, err := tf.GetStateOutputVariables(ctx, infrastructure.TerraformOutputKeyRouterID)
```

Comment:
This is an interesting choice to use the TF methods to get the router ID and I want to make sure that I understand the implications.

I am concerned with the following scenario
- creating a shoot with invalid credentials and want to delete it afterwards. In this case the state may or may not be there. The router variable will also certainly not be there. From a glance I think that your calls later e.g. `router, err := client.GetRouterByID(routerID)`  would permanently block deletion

I also think that the condition at line 103 `DoIf(configExists)` is not entirely correct. The config may exists but a routerID not necessarily. 

Maybe we should use the infrastructureStatus for the routerID ? That way you would know whether there is a router in the first place and you can use the ID directly instead of making the lookup

Reply:
It's similar to what we do on gcp https://github.com/gardener/gardener-extension-provider-gcp/blob/eb851f716e45336b486f3aaf46268859de2adecb/pkg/controller/infrastructure/actuator_delete.go#L62

Reply:
It's all done in the `ExtractTerraformState` function.

Reply:
If there is no routerID I will assume that there are no routes to delete and go on with the tf.Destory.